lib/Dialect/AIE/Transforms/AIEAssignLockIDs.cpp

```diff
@@ -40,50 +39,63 @@ struct AIEAssignLockIDsPass : AIEAssignLockIDsBase<AIEAssignLockIDsPass> {
       }
     }
 
-    // The second pass scans for locks with no lockIDs and assigns locks.
-    for (auto lock : device.getOps<LockOp>()) {
-      Operation *lock_tile = lock.getTile().getDefiningOp();
-      if (!lock.getLockID().has_value()) {
-        if (tileToLastID.find(lock_tile) == tileToLastID.end()) {
-          // If the tile operation corresponding to the lock does not exist in
-          // the data structure, initialize the lockID with 0 with an empty set.
-          tileToLastID[lock_tile].first = 0;
-        } else if (tileToLastID[lock_tile].first < 15) {
-          // If the tile operation of the lock exists, the potential lockID is
-          // checked with the set containing occupied lockIDs until a lockID
-          // that is free is found.
-          int potential_ID = tileToLastID[lock_tile].first;
-          while (true) {
-            if (tileToLastID[lock_tile].second.find(potential_ID) !=
-                tileToLastID[lock_tile].second.end())
-              potential_ID++;
-            else
-              break;
+    // For each tile, ensure that all its locks have unique and valid IDs.
+    for (auto &&[tileOp, locks] : tileToLockOps) {
+
+      const auto locksPerTile =
+          getTargetModel(tileOp).getNumLocks(tileOp.getCol(), tileOp.getRow());
+
+      // All lock IDs which are assigned to locks before this pass.
+      DenseSet<int> idsPreAssigned;
+
+      // All lock ops which do not have IDs before this pass.
+      SmallVector<LockOp> idsNotAssigned;
+
+      for (auto &&lockOp : locks) {
+        if (lockOp.getLockID().has_value()) {
+          uint32_t lockID = lockOp.getLockID().value();
+
+          assert(lockID < locksPerTile &&
+                 "This should be checked in the lock op verifier");
+
+          // Duplicate preassigned lock ID:
+          if (idsPreAssigned.find(lockID) != idsPreAssigned.end()) {
+            tileOp->emitOpError("has multiple locks with ID ") << lockID << '.';
+            return signalPassFailure();
           }
-          tileToLastID[lock_tile].first = potential_ID;
+          idsPreAssigned.insert(lockID);
         } else {
-          lock->emitError() << "Exceeded the number of unique LockIDs";
-          return;
+          idsNotAssigned.push_back(lockOp);
         }
+      }
+
+      assert(idsPreAssigned.size() + idsNotAssigned.size() == locks.size());
 
-        // The lockID is assigned and is stored in the set.
-        lock->setAttr("lockID", rewriter.getI32IntegerAttr(
-                                    tileToLastID[lock_tile].first));
-        tileToLastID[lock_tile].second.insert(tileToLastID[lock_tile].first);
+      uint32_t nxtId = 0;
+      for (auto &&lockOp : idsNotAssigned) {
+        while (nxtId < locksPerTile &&
+               (idsPreAssigned.find(nxtId) != idsPreAssigned.end())) {
+          ++nxtId;
+        }
+        if (nxtId == locksPerTile) {
+          tileOp->emitOpError("can have a maximum of ")
```

Comment:
Would make more sense if this was an error on the lock op, with a note pointing to the tile op.

Reply:
Updated. Now looks something like

within split at X.mlir:124 offset :24:10: error: 'AIE.lock' op was not allocated a lock.
  %l16 = AIE.lock(%tMemTile)
         ^
within split at X.mlir:124 offset :24:10: note: see current operation: %17 = "AIE.lock"(%0) : (index) -> index
within split at X.mlir:124 offset :7:15: note: tile only has 16 locks available.
  %tMemTile = AIE.tile(4,4)
              ^